cpp/src/prims/detail/nbr_intersection.cuh

```diff
@@ -392,31 +432,82 @@ struct copy_intersecting_nbrs_and_update_intersection_size_t {
         static_cast<edge_t>(second_element_offsets[idx + 1] - second_element_offsets[idx]);
       indices1 = second_element_indices.begin() + second_element_offsets[idx];
+
+      local_edge_offset1 = second_element_offsets[idx];
+      if constexpr (!std::is_same_v<edge_property_value_t, thrust::nullopt_t>) {
+        property1 = second_element_properties.begin() + local_edge_offset1;
+      }
     }
 
     // FIXME: this can lead to thread-divergence with a mix of high-degree and low-degree
     // vertices in a single warp (better optimize if this becomes a performance
     // bottleneck)
 
-    auto it =
-      thrust::set_intersection(thrust::seq,
-                               indices0,
-                               indices0 + local_degree0,
-                               indices1,
-                               indices1 + local_degree1,
-                               nbr_intersection_indices.begin() + nbr_intersection_offsets[i]);
+    auto nbr_intersection_first = nbr_intersection_indices.begin() + nbr_intersection_offsets[i];
+
+    auto nbr_intersection_last = thrust::set_intersection(thrust::seq,
+                                                          indices0,
+                                                          indices0 + local_degree0,
+                                                          indices1,
+                                                          indices1 + local_degree1,
+                                                          nbr_intersection_first);
     thrust::fill(thrust::seq,
-                 it,
+                 nbr_intersection_last,
                  nbr_intersection_indices.begin() + nbr_intersection_offsets[i + 1],
                  invalid_id);
 
-    return static_cast<size_t>(
-      thrust::distance(nbr_intersection_indices.begin() + nbr_intersection_offsets[i], it));
+    auto insection_size =
+      static_cast<size_t>(thrust::distance(nbr_intersection_first, nbr_intersection_last));
+    if constexpr (!std::is_same_v<edge_property_value_t, thrust::nullopt_t>) {
+      auto ip0_start = nbr_intersection_properties0.begin() + nbr_intersection_offsets[i];
+
+      // copy edge properties from first vertex to common neighbors
+      thrust::lower_bound(thrust::seq,
+                          indices0,
+                          indices0 + local_degree0,
+                          nbr_intersection_first,
+                          nbr_intersection_last,
+                          nbr_intersection_idx_buffer.begin() +
+                            nbr_intersection_offsets[i],  // ip0_start,  // indices
+                          thrust::less<int>());
+
+      thrust::transform(
+        thrust::seq,
+        nbr_intersection_idx_buffer.begin() + nbr_intersection_offsets[i],
+        nbr_intersection_idx_buffer.begin() + nbr_intersection_offsets[i] + insection_size,
+        ip0_start,
+        [property0] __device__(auto idx) { return property0[static_cast<int>(idx)]; });
```

Comment:
Do we really need this temporary index buffer? Can't we do something like the following?

```
thrust::transform(thrust::seq,
  nbr_intersection_first,
  nbr_intersection_last,
  ip0_start,
  [...]__device__(...) {
    auto idx = thrust::lower_bound();
  return property0idx]; });
```